static/js/silhouette.js

```diff
@@ -0,0 +1,98 @@
+'use strict'
+
+/**
+ * Calculate euclidean distance between two points
+ * @function euclidean_dist
+ * @param {array} x - coordinate of point x
+ * @param {array} y - coordinate of point y
+ * @return {number} euclidean distance between x and y
+ */
+function euclidean_dist(x, y) {
+	// check x, y
+	let sum = 0;
+	for (let i = 0; i < x.length; i++) {
+		sum += (x[i] - y[i]) ** 2;
+	}
+	return Math.sqrt(sum);
+}
+
+/**
+ * @function pairwise_dist
+ * @param {Array} x - the input data array
+ * @return {Array} distance matrix of the given data
+ */
+//TODO: add metrics='euclidean_dist'
+function pairwise_dist(x) {
+	let d = Array(x.length).fill().map(()=>Array(x.length).fill());
+	for (let i = 0; i < x.length; i++) {
+		for (let j = 0; j < x.length; j++) {
+			if (j < i) {
+				d[i][j] = d[j][i];
+			} else {
+				d[i][j] = euclidean_dist(x[i], x[j]);
+			}
+		}
+	}
+	return d;
+}
```

Comment:
I am thinking that the current 2D array implementation may not be the most efficient. Because the distance is not directional, the upper triangle and the lower triangle are always the same. Keeping them both in the memory is not ideal. In scipy there is a "condensed" form of distance matrix (see here: https://docs.scipy.org/doc/scipy/reference/spatial.distance.html) which you can perhaps refer to.